Ticket 1035 - Simple Solution

rprospero opened the pull request:
This fixes ticket #1035 by ensuring that the product of the form and structure factor takes precedence over the addition of other models, as one would expect from order of operations.

There do not appear to be any unit tests which check combined models, so this code has only been tested for regressions manually.  To minimize the chance of introducing new bugs, this pull request seeks to simplify and minimize the code in load_model_info.  The new parser is almost 40% shorter, uses only a single level of conditionals, and makes the order of operations explicit.

As this ticket solves the same ticket as pull request #60, only one of these two should be accepted.

pkienzle commented:
I suggest this version, which gives cleaner code than #60.

I think you need to test for ```@``` last so that ```a*b@c``` is parsed as ```a*(b@c)```.  We don't define ```ER``` for ```a*b```, so ```(a*b)@c``` isn't valid.

rprospero commented:
I've added the fix for Python 3 and tested it again in Python 2.  I've also added the provided unit test, in addition to another series of tests to confirm that the order of the models in the definition doesn't change the outcome.

butlerpd commented:
closes #1035
